plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java

```diff
@@ -0,0 +1,108 @@
+import io.trino.spi.type.CharType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.VarcharType;
+import oracle.jdbc.OracleTypes;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argument;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argumentCount;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.call;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.expression;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.functionName;
+import static io.trino.plugin.oracle.OracleClient.ORACLE_CHAR_MAX_CHARS;
+import static io.trino.plugin.oracle.OracleClient.ORACLE_VARCHAR2_MAX_CHARS;
+import static io.trino.spi.expression.StandardFunctions.CAST_FUNCTION_NAME;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class RewriteCast
+        implements ProjectFunctionRule<JdbcExpression, ParameterizedExpression>
+{
+    private static final Capture<ConnectorExpression> VALUE = newCapture();
+    private final BiFunction<ConnectorSession, Type, WriteMapping> toWriteMapping;
+
+    public RewriteCast(BiFunction<ConnectorSession, Type, WriteMapping> toWriteMapping)
+    {
+        this.toWriteMapping = requireNonNull(toWriteMapping, "toWriteMapping is null");
+    }
+
+    @Override
+    public Pattern<Call> getPattern()
+    {
+        return call()
+                .with(functionName().equalTo(CAST_FUNCTION_NAME))
+                .with(argumentCount().equalTo(1))
+                .with(argument(0).matching(expression().capturedAs(VALUE)));
+    }
+
+    @Override
+    public Optional<JdbcExpression> rewrite(ConnectorExpression projectionExpression, Captures captures, RewriteContext<ParameterizedExpression> context)
+    {
+        Call call = (Call) projectionExpression;
+        Type trinoType = call.getType();
+        ConnectorExpression capturedValue = captures.get(VALUE);
+
+        Optional<ParameterizedExpression> value = context.rewriteExpression(capturedValue);
+        if (value.isEmpty()) {
+            // if argument is a call chain that can't be rewritten, then we can't push it down
+            return Optional.empty();
+        }
+        Optional<JdbcTypeHandle> targetTypeHandle = toJdbcTypeHandle(trinoType);
+        if (targetTypeHandle.isEmpty()) {
+            return Optional.empty();
+        }
+
+        String targetType = toWriteMapping.apply(context.getSession(), trinoType).getDataType();
+
+        return Optional.of(new JdbcExpression(
+                format("CAST(%s AS %s)", value.get().expression(), targetType),
+                value.get().parameters(),
+                targetTypeHandle.get()));
+    }
+
+    private static Optional<JdbcTypeHandle> toJdbcTypeHandle(Type type)
```

Comment:
This method is the meat of this PR. 
And this is a main opportunity to improve for function pushdown.